### ISSUE TITLE
Create alias test.k8s.io/kubernetes-e2e/12345 for GCS artifacts?

### ISSUE THREAD

**Issue by ihmccreery** [kubernetes/kubernetes#20161]
I think it might be worth creating some kind of alias for people to easily find test results.  @thockin and I have discussed how Jenkins links are non-permanent, and of course they are also not public.  It would be easy to print out a statement at the top of a Jenkins run that says something like:

```
Test results will be available at
  http://test.k8s.io/kubernetes-e2e/12345
once this test completes
```

and that second line could be what people could post to issues, etc.

cc @kubernetes/goog-testing

**Comment by thockin** [comment]
I can do this after some meetings today.  If you can give me teh transform
to apply I'll drop it into the redirector

**Comment by spxtr** [comment]
It currently says this at the very bottom of the test logs.

```
+ echo -e '\n\n\n*** View logs and artifacts at https://storage.cloud.google.com/kubernetes-jenkins/logs/kubernetes-e2e-gce/10539 ***\n\n'



*** View logs and artifacts at https://storage.cloud.google.com/kubernetes-jenkins/logs/kubernetes-e2e-gce/10539 ***


No emails were triggered.
Finished: SUCCESS
```

**Comment by thockin** [comment]
Ike,

Can you summarize the various discussions?  Given a PR #, what URL should I
redirect to?

**Comment by ixdy** [comment]
FYI kubernetes/contrib/pull/404 just merged, which will link the SQ status page to the correct GCS storage browser urls.

**Comment by ihmccreery** [comment]
**For CI**: we should redirect `http://ci-test.k8s.io/some-kind-of-test-suite/12345` to `https://storage.cloud.google.com/kubernetes-jenkins/logs/some-kind-of-test-suite/12345`; (that's run `12345` of `some-kind-of-test-suite`, e.g. `kubernetes-e2e-gce`).

**For PRs**: per https://github.com/kubernetes/kubernetes/pull/20166, given a PR number, maybe we can redirect `http://pr-test.k8s.io/12345` to `https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/12345`?

@ixdy does that sound right to you?

**Comment by ixdy** [comment]
Yep, sounds correct.

**Comment by ixdy** [comment]
That's basically correct, but we should probably be consistent with the base URL - e.g. either `https://storage.cloud.google.com/` or `https://console.developers.google.com/storage/browser/`. The former seems to redirect to the latter.

**Comment by ihmccreery** [comment]
Fewer redirects seems reasonable to me; let's stick with `console.developers....`?

**Comment by ixdy** [comment]
OK, I just found one subtle difference between these two, mostly when linking to an object in the bucket (rather than a path):
https://storage.cloud.google.com/kubernetes-jenkins/pr-logs/pull/20143/kubernetes-pull-build-test-e2e-gce/25287/build-log.txt works.
https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/20143/kubernetes-pull-build-test-e2e-gce/25287/build-log.txt doesn't, because it tries to treat "build-log.txt" as a path, rather than the name of an object.

Arguably this is a bug in the developer console, but we should probably use the "storage.cloud.google.com" URL to work around it.

**Comment by thockin** [comment]
Done.  PTAL and close this issue.

**Comment by ihmccreery** [comment]
`ci-test.k8s.io/<suite>/<run>` and `pr-test.k8s.io/<pr>`.  Thanks @thockin!

**Comment by ixdy** [comment]
We need to revisit this. Apparently the storage.cloud.google.com redirects to the storage browser were an unintended undocumented feature, and it's been disabled. But we can't just redirect everything to `https://console.developers.google.com/storage/browser`, either, for the reason I noted in https://github.com/kubernetes/kubernetes/issues/20161#issuecomment-175296750.

@thockin how advanced is the redirect logic? Can we redirect a URL ending in `/` to `https://console.developers.google.com/storage/browser` and anything else to `https://storage.cloud.google.com/`? It's not perfect, but it'd probably mostly work.

**Comment by thockin** [comment]
I can do semi-arbitrary regex-based redirects.  That includes sub-paths.  Can you be more specific about what we want and for which of these two URLs?

**Comment by ixdy** [comment]
OK, great. What I think we want:
* `ci-test.k8s.io/(.*)`
  * if `$1` ends in `/`, redirect to `https://console.developers.google.com/storage/browser/kubernetes-jenkins/logs/$1`
  * else redirect to `https://storage.cloud.google.com/kubernetes-jenkins/logs/$1`
* `pr-test.k8s.io/(.*)`
  * if `$1` ends in `/`, redirect to `https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/$1`
  * else redirect to `https://storage.cloud.google.com/kubernetes-jenkins/pr-logs/pull/$1`

The intended effect here is that requesting a file downloads/shows the file, whereas requesting a "directory" opens the storage browser.

**Comment by thockin** [comment]
puke.  The trailing / is strictly optional in a URL, isn't it?  Is there no better pattern to match or backend to hit that can unify or maybe give  permissions to browse https://storage.cloud.google.com/kubernetes-jenkins/pr-logs/pull/ directories?

**Comment by ixdy** [comment]
The only other way I can think of to determine whether to redirect to the browser or to the GCS frontend (storage.cloud.google.com) is to try the path on the GCS frontend and see if it 404s, but that's probably out of scope for the redirector.

I also just opened kubernetes/contrib#734 to suggest creating a web interface to eliminate our dependency on the storage browser entirely.

**Comment by thockin** [comment]
Is there no way to make storage.cloud.google.com/kubernetes-jenkins/{logs,pr-logs} browseable so that directory URLs work?  It looks like it is possible, but it still asks you login to google to browse.  If you go thru the http://storage.googleapis.com URL a "browse" gets you XML data which isn't very friendly either.  I feel like there's an answer down this path, though...

What if we looked for known file-extensions in the redirects ? (log, json, xml, yaml, txt, ...)?

**Comment by thockin** [comment]
I have not yet found a better answer.  Thoughts on resolving this?  I can implement the trailing / very easily, and if we're confident that the links we generate fit this pattern, maybe we should just do it?

**Comment by ixdy** [comment]
We can make sure that links we generate follow this pattern. It means that someone manually creating a path (like http://pr-test.k8s.io/1234/) will need to add the trailing slash or it won't work, which is annoying, but better than the current fully-broken state.

**Comment by thockin** [comment]
I have made this change in the URL redirector.
